compiler/rustc_hir_typeck/src/fn_ctxt/arg_matrix.rs

```diff
@@ -376,6 +376,13 @@ impl<'tcx> ArgMatrix<'tcx> {
             };
         }
 
+        // sort errors with same type by the order they appear in the source
+        // so that suggestion will be handled properly, see #112507
+        errors.sort_by(|a, b| match (a, b) {
+            (Error::Missing(i), Error::Missing(j)) => i.cmp(j),
```

Comment:
I think you could probably just derive `Ord` on `Error` and just use `.sort()`?

Reply:
we need to customize `PartialOrd` for `Error` and I think only ordering on `Missing` and `Extra` is enough(which will involve adding or removing arguments).

Reply:
I don't think it hurts to just totally sort the list. I think right now the sort implementation is not actually valid because it returns Equal for other elements 😅

Reply:
yes, I just validated with some playground code, it won't sort in some cases, let me fix it 👍

Reply:
But actually, this implementation is enough, because it makes sure the **consecutive** `Extra` and `Missing` ordered 😆, which is the scenario that makes sub-suggestions buggy.

Anyway, let's make it all ordered.